runtime/webhookserver/webhooktunnel.go

```diff
@@ -0,0 +1,94 @@
+package webhookserver
+
+import (
+	"errors"
+	"net/http"
+	"sync"
+
+	"github.com/taskcluster/slugid-go/slugid"
+	"github.com/taskcluster/taskcluster-client-go"
+	"github.com/taskcluster/taskcluster-client-go/auth"
+	"github.com/taskcluster/webhooktunnel/whclient"
+)
+
+// WebhookTunnel wraps a whclient instance as a WebHookServer
+type WebhookTunnel struct {
+	m        sync.RWMutex
+	handlers map[string]http.Handler
+	// url is assigned when the client connects to the proxy
+	client *whclient.Client
+}
+
+// NewWebhookTunnel returns a pointer to a new WebhookTunnel instance
+func NewWebhookTunnel(credentials *tcclient.Credentials) (*WebhookTunnel, error) {
+	configurer := func() (whclient.Config, error) {
+		authClient := auth.New(credentials)
+		whresp, err := authClient.WebhooktunnelToken()
+		if err != nil {
+			return whclient.Config{}, errors.New("could not get token from tc-auth")
+		}
+
+		return whclient.Config{
+			ID:        whresp.TunnelID,
+			ProxyAddr: whresp.ProxyURL,
+			Token:     whresp.Token,
+		}, nil
+	}
+
+	client, err := whclient.New(configurer)
+	if err == whclient.ErrAuthFailed {
+		client, err = whclient.New(configurer)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	wt := &WebhookTunnel{
+		handlers: make(map[string]http.Handler),
+		client:   client,
+	}
+
+	go func() {
+		_ = http.Serve(wt.client, http.HandlerFunc(wt.handle))
```

Comment:
Can we leave a TODO that an error here should be reported to sentry through monitor, when we finally add that as well...

I'm thinking you can do that in a follow up PR... if you have time...